Boards: STM32H750B-DK: Introduce App in Ext Flash variant

JarmouniA opened the pull request:
Introduce a board variant for storing an app in external QSPI NOR Flash and chainloading it with MCUboot (which is placed in internal Flash) to be executed in place (XiP).

The ext Flash memory partition also allows for using the Swap-using-offset & swap-using-move MCUboot image upgrade algorithms, eventhough it's not optimal as it has same-sized slots.

The new variant requires a board DT overlay on the MCUboot side (https://github.com/mcu-tools/mcuboot/pull/2479) to set the internal Flash & Flash controller as the chosen 'zephyr,flash' & 'zephyr,flash-controller'. This way, no overlay will be needed at the Zephyr user app level.

```
west build -p always -b stm32h750b_dk/stm32h750xx/ext_flash_app samples/modules/lvgl/demos --sysbuild
```

Board doc preview https://builds.zephyrproject.io/zephyr/pr/97037/docs/boards/st/stm32h750b_dk/doc/index.html

JarmouniA committed:
boards: stm32h750b_dk: introduce board variant for apps in ext Flash

Introduce a board variant for storing an app in external QSPI NOR Flash
and chainloading it with MCUboot (placed in internal Flash) to be executed
in place (uses MCUboot's swap-using-offset mode, set via Sysbuild).

The new variant requires on MCUboot app side: a board/SoC DT overlay to
set the internal Flash & Flash controller as the chosen 'zephyr,flash' &
'zephyr,flash-controller', and CONFIG_STM32_MEMMAP=y (this was upstreamed
to MCUboot.)

Change external Flash memory attribute "zephyr,memory-attr" to
DT_MEM_ARM_MPU_FLASH, since ATTR_MPU_IO (corresponds to
"DEVICE_NON_SHAREABLE & P_RW_U_NA") is actually for memory-mapped devices
(with registers...) and was causing access issues with apps like LVGL that
would freeze the target. Also, change the size of the external flash
"zephyr,memory-region" node, since we only need to define accesss rights
to the memory space actually used by the external flash chips, not the
total addresssable QSPI memory.

The 2nd QSPI NOR contoller node was removed, since it does not reflect the
actual hardware layout, that has only one controller and the 2 NOR flash
chips are connected to it in parallel with a shift register.

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

erwango reviewed (commented):
@JarmouniA I was wondering if snippets could be used instead of adding a target each time we need to introduce an ext_mem app. Did you look at this possibility ?
Just thinking out loud, maybe the combination with sysbuild doesn't work great.

JarmouniA committed:
samples: sysbuild: with_mcuboot: add stm32h750b_dk to CI platforms

Add stm32h750b_dk/stm32h750xx/ext_flash_app target to CI platforms

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA committed:
tests: boot: test_mcuboot: add stm32h750b_dk to CI platforms

Add stm32h750b_dk/stm32h750xx/ext_flash_app target to CI platforms.

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

erwango reviewed (changes requested):
Nice!

Though, I think we can get rid of the mcuboot overlay by using `stm32h750b_dk` as the mcuboot target. This way, mcuboot is a usual zephyr application running on internal flash.
(Note also that `zephyr,code-partition = &boot_partition;` is already defined in mcuboot app.overlay)

To do this, you probably need to add a `sysbuild.cmake` file to the board, stating:
```
if(SB_CONFIG_BOOTLOADER_MCUBOOT)
  set_target_properties(mcuboot PROPERTIES BOARD stm32h750b_dk/stm32h750xx)
endif()
```
and no mcuboot dependency anymore.

Can you have a try and report back (somehow I can't find any stm32h750b_dk here anymore to test it myself) ?

erwango commented:
> The MCUboot partitions (slot0, slot1...) are not relevant to the base variant, which is supposed to be used with a regular Zephyr app (MCUboot is not a regular app, it's a bootloader) in internal Flash & external Flash is used for storage or whatever,

We already have this case on other boards where "main" variant .dts provides slot0/slot1 on internal flash and this "main" variant could be used either as a "direct" zephyr application or chainloaded behind mcuboot.

Difference here is that we need a variant to be able to specify different `zephyr,flash` and `zephyr,flash-controller`, hence the need for a dedicated variant (ext_flash_app). With your proposal, we don't need to provide a .conf file anymore for the MCUBoot target, let's go one step further and get rid of the .overlay and dependency on MCUBoot completely.

Also note that these boards (based on h750, h7r/s) are a bit specific in the way that they have a quite limited internal flash and are meant to be used with a ext_flash, from which the main application should run.
Having the main variant running on internal flash is convenient in order to be able to run upstream zephyr samples and tests, but proposing the external flash as an extra large storage slot is only interesting for validation purpose, not as a real final app. If co-existence of this storage slot with slot0/slot1 partition on the main variant raises an issue, then it should be defined as an overlay in tests/samples that actually need it.

Last point about interest of the ext_flash_app variant is that it could be used to specify configurations that would not be supported on the main variant running on the internal flash (thinking about video use cases for instance). But I need to verify this point.

erwango commented on boards/st/stm32h750b_dk/stm32h750b_dk_stm32h750xx_ext_flash_app.dts:
```diff
@@ -0,0 +1,69 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			slot0_partition: partition@0 {
```
Did you try with `CONFIG_XIP=n` ? (required when RAM_LOAD is used)

aescolar commented:
Note: The last removal of reviewers was a side-effect of the 15 reviewer limit having being surpassed. And not something I did deliberately.

nordicjm commented on boards/st/stm32h750b_dk/board.cmake:
```diff
@@ -1,7 +1,8 @@
-if(CONFIG_STM32_MEMMAP)
+if(CONFIG_STM32_MEMMAP OR CONFIG_BOOTLOADER_MCUBOOT)
 board_runner_args(stm32cubeprogrammer "--port=swd" "--reset-mode=hw")
 board_runner_args(stm32cubeprogrammer "--extload=MT25TL01G_STM32H750B-DISCO.stldr")
```
cmake indent is 2 spaces

nordicjm commented on boards/st/stm32h750b_dk/stm32h750b_dk_stm32h750xx_ext_flash_app_defconfig:
```diff
@@ -0,0 +1,20 @@
+
+CONFIG_SERIAL=y
+
+# console
```
capital first letter for comments

nordicjm commented on boards/st/stm32h750b_dk/Kconfig.defconfig:
```diff
@@ -14,4 +14,21 @@ config MEMC
+# So MCUboot can access ext Flash in memory-mapped mode
+# to read app slots, start the app, or copy it to RAM.
+config STM32_MEMMAP
+	default y
```
this should be in a board Kconfig fragment file in the mcuboot repo

JarmouniA replied:
Wouldn't that prevent MCUboot's `boot/zephyr/prj.conf` from being applied?

JarmouniA commented:
@tpambor You should take into consideration the original default of FLASH_BASE_ADDRESS  as well,
https://github.com/zephyrproject-rtos/zephyr/blob/fd51dde8f5cabb557bd5a34fa2c44b374b2647fc/arch/Kconfig#L260 
which is what applies in this PR.

JarmouniA commented:
@tpambor You should take into consideration the other STM32 default of FLASH_BASE_ADDRESS as well,
https://github.com/zephyrproject-rtos/zephyr/blob/fd51dde8f5cabb557bd5a34fa2c44b374b2647fc/soc/st/stm32/Kconfig.defconfig#L104 
which is what applies in this PR, because none of the conditions for the other default is true.

But I agree that the condition `$(DT_FLASH_CTRL_IS_{Q,O,X}SPI)` should be used instead of `$(DT_FLASH_PARENT_IS_{Q,O,X}SPI)`, but it should be done in a separate PR. 

`ranges` is needed for another reason, see https://github.com/zephyrproject-rtos/zephyr/pull/97413

erwango commented on boards/st/stm32h750b_dk/stm32h750b_dk.dts:
```diff
@@ -5,288 +5,29 @@
-		size = <DT_SIZE_M(512)>; /* 512 Mbits */
-		qspi-max-frequency = <72000000>;
-		status = "okay";
-	};
```
Suppression of this bank should be at least documented and explain in the commit message.

erwango commented on boards/st/stm32h750b_dk/stm32h750b_dk-common.dtsi:
```diff
@@ -0,0 +1,266 @@
+
+	ext_flash_mem: memory@90000000 {
+		compatible = "zephyr,memory-region";
+		reg = <0x90000000 DT_SIZE_M(128)>; /* 128MB */
```
Change of size would also need to be detailled (maybe linked to bank2 removal ? Is it safe or justified ?)

erwango reviewed (changes requested):
Some more comments

JarmouniA commented on boards/st/stm32h750b_dk/stm32h750b_dk-common.dtsi:
```diff
@@ -0,0 +1,266 @@
+		reg = <0x90000000 DT_SIZE_M(128)>; /* 128MB */
+		zephyr,memory-region = "EXT_FLASH";
+		zephyr,memory-attr = <DT_MEM_ARM_MPU_FLASH>;
+	};
```
Change detailed in the commit message.

JarmouniA commented:
@etienne-lms Could you take a look, thanks!

etienne-lms commented on boards/st/stm32h750b_dk/stm32h750b_dk_stm32h750xx_ext_flash_app.yaml:
```diff
@@ -0,0 +1,19 @@
+  - zephyr
+sysbuild: true
+ram: 1024
+flash: 61439  # size in kB of 1 app slot minus MCUboot header size (1KB)
```
Also count the MCUboot trailer? less than 1KB if I'm right.

JarmouniA commented on boards/st/stm32h750b_dk/Kconfig.sysbuild:
```diff
@@ -0,0 +1,18 @@
+
+choice MCUBOOT_MODE
+	default MCUBOOT_MODE_DIRECT_XIP
+endchoice
```
@erwango swap-using-move and swap-using-offset do work (even if the partitioning is not optimal, the slots are equal size).
The app with DIRECT_XIP starts faster, upgrading the image is nearly instantaneous (no copying needed, esp. when the image is split between 2 Flash chips, also means less flash wear), and there is no "reset in the middle of an image swap" issue to deal with.

maass-hamburg replied:
> @erwango swap-using-move and swap-using-offset do work (even if the partitioning is not optimal, the slots are equal size). The app with DIRECT_XIP starts faster, upgrading the image is nearly instantaneous (no copying needed, esp. when the image is split between 2 Flash chips, also means less flash wear), and there is no "reset in the middle of an image swap" issue to deal with.

Slower but working for more samples, for a dev board this is fine. And for the samples that support direct xip, there we could provide a either provide documentation how to improve speed or a extra conf file.

JarmouniA replied:
> And for the samples that support direct xip

All samples, except those doing image upgrade stuff, do not care about the mode, `west build --sysbuild.. & flash` builds then flashes the slot0 image by default.

maass-hamburg replied:
> > And for the samples that support direct xip
> 
> All samples, except those doing image upgrade stuff, do not care about the mode, `west build --sysbuild.. & flash` builds then flashes the slot0 image by default.

yes and they also don't do move between the 2 slots, so they shouldn't be slower or?

JarmouniA replied:
> set CONFIG_BOOT_MAX_IMG_SECTORS higher, as it has to be increased for a bigger slot size.
And CONFIG_BOOT_MAX_IMG_SECTORS_AUTO is also then deactivated

Not needed, erase-block-size is set in the ext Flash node.

> I would prefer the use of MCUBOOT_MODE_OVERWRITE_ONLY

It does not work as well, same error as Swap modes.

Will push an update with SINGLE_APP as default.

maass-hamburg replied:
for this board as it not working with the swap modes, yes

We can try to fix it maybe later, after the release

maass-hamburg commented on boards/st/stm32h750b_dk/stm32h750b_dk-common.dtsi:
```diff
@@ -0,0 +1,266 @@
+			compatible = "soc-nv-flash";
+			reg = <0x0 DT_SIZE_M(128)>; /* 128MB total in dual-flash mode */
+			write-block-size = <1>; /* 1byte=4bits*2 in dual-flash mode */
+			erase-block-size = <DT_SIZE_K(128)>; /* 64KB*2 in dual-flash mode */
```
This should be the lowest supposed
```suggestion
			erase-block-size = <DT_SIZE_K(8)>; /* 4KB*2 in dual-flash mode */
```

JarmouniA replied:
Actually, great catch, it fixed the Swap modes 🥳

JarmouniA committed:
samples: mgmt: hawkbit: add stm32h750b_dk to CI platforms

Add stm32h750b_dk/stm32h750xx/ext_flash_app target to CI platforms of
the Hawkbit sample.

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

maass-hamburg reviewed (approved):
As things should be 😌
